File: Client/src/main/java/com/rspsi/editor/model/WorldTileSource.java
package com.rspsi.editor.model;

import java.util.Objects;

/** A decoded tile plus the cache height opcode semantics that produced it. */
public record WorldTileSource(TileSnapshot snapshot, TerrainHeightSource heightSource) {
    public WorldTileSource {
        snapshot = Objects.requireNonNull(snapshot, "snapshot");
        heightSource = Objects.requireNonNull(heightSource, "heightSource");
    }
}

File: Client/src/main/kotlin/com/rspsi/editor/model/WorldTileSource.kt
package com.rspsi.editor.model

/**
 * A decoded tile plus the cache-height provenance that produced it.
 *
 * Kept as a real JVM record so Java callers retain component accessors such as
 * snapshot() and heightSource() while Kotlin gains native data-class ergonomics.
 */
@JvmRecord
data class WorldTileSource(
    val snapshot: TileSnapshot,
    val heightSource: TerrainHeightSource,
) {
    override fun toString(): String =
        "WorldTileSource[snapshot=$snapshot, heightSource=$heightSource]"
}
